### PR TITLE
Add implementation of pseudo-inverse

### PR DESCRIPTION
**Pull request by jlecoeur**
The implementation is based on this publication:
Courrieu, P. (2008). Fast Computation of Moore-Penrose Inverse Matrices, 8(2), 25–29. http://arxiv.org/abs/0804.4809
It is a fully templated implementation to guaranty type correctness.

### CODE REVIEW

**Comment by RomanBapst** [comment]
@jlecoeur Cool stuff. I'll need a bit more time to review this. Great that you have already included tests!

**Comment by MaEtUgR** [comment]
Nice! Seems there are not enough tests yet: "Coverage decreased"

**Comment by MaEtUgR** [comment]
Is there no overlap with these?
https://github.com/PX4/Matrix/blob/215203fc6fcb87d79a1ad8ef390fd243e6306d08/matrix/SquareMatrix.hpp#L161-L283
https://github.com/PX4/Matrix/blob/215203fc6fcb87d79a1ad8ef390fd243e6306d08/matrix/SquareMatrix.hpp#L305-L336
It might be worth replacing or reusing them.

**Comment by jlecoeur** [comment]
@MaEtUgR I do not know how to please coveralls here. It complains that two lines in the implementation detail are not covered. This is indeed a case that should never happen. I tried to cover it in the tests (see the last test) but it did not help.

> Is there no overlap with these?

Not for `inv`: `geninv` is defined for non-square matrices while `inv` is defined for square matrices. If you pass a square matrix to `geninv`, it will fall back to `inv`.

For `cholesky`, maybe. The main difference is that the new one provides the rank. Do you suggest to replace the existing one with the new one?

File: matrix/PseudoInverse.hxx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+SquareMatrix<Type, N> full_rank_cholesky(const SquareMatrix<Type, N> & A,
+                               size_t& rank)
+{
+    // Compute 
+    // dA = np.diag(A)
+    // tol = np.min(dA[dA > 0]) * 1e-9
+    Vector<Type, N> d = A.diag();
+    Type tol = d.max();
+    for (size_t k = 0; k < N; k++) {
+        if (d(k) > 0 && (d(k) < tol)) {
```
[the remark this answers is not in the document]
Comment (jlecoeur) [reply matrix/PseudoInverse.hxx]: Yes, I added brackets for extra clarity.

**Comment by jkflying** [comment]
Looks cool, I'll try to do an in-depth review on Monday.

It looks like the decreased code coverage is due to the (illegal, IIRC) error cases not being executed. Maybe there's another way to phrase the check so that we don't end up with this happening, or maybe we can force it to happen in tests by setting up sections of the problem separately?

Also, it would be good to do some performance testing of this against something like Eigen to know if we still have some easy gains anywhere.

**Comment by jlecoeur** [comment]
> If you pass a square matrix to geninv, it will fall back to inv.

This is actually a very bad idea because there is good chance that the matrix is not invertible. I don't know what I was thinking! I will remove the template specialization.

File: matrix/PseudoInverse.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+template<typename Type, size_t M, size_t N>
+Matrix<Type, N, M> geninv(const Matrix<Type, M, N> & G)
+{
+    size_t rank;
+    if (M <= N) {
+        SquareMatrix<Type, M> A = G * G.transpose();
+        SquareMatrix<Type, M> L = full_rank_cholesky(A, rank);
+
+        return Geninv_impl<Type, M, N, M>::geninv_M(G, L, rank);
```
Comment (bresch) [inline matrix/PseudoInverse.hpp]:
```suggestion
        return GenInvImpl<Type, M, N, M>::genInvUnderdetermined(G, L, rank);
```

File: matrix/PseudoInverse.hpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    } else {
+        SquareMatrix<Type, N> A = G.transpose() * G;
+        SquareMatrix<Type, N> L = full_rank_cholesky(A, rank);
+
+        return Geninv_impl<Type, M, N, N>::geninv_N(G, L, rank);
```
Comment (bresch) [inline matrix/PseudoInverse.hpp]:
```suggestion
        return GenInvImpl<Type, M, N, N>::genInvOverdetermined(G, L, rank);
```

File: matrix/PseudoInverse.hpp (hunk 2 above)
Comment (jlecoeur) [reply matrix/PseudoInverse.hpp]: Much better!

**Comment by jlecoeur** [comment]
Here are runtimes on a Pixhawk 4:
geninv_bench_2
and on a Pixhawk 1:
geninv_bench_F4

File: matrix/PseudoInverse.hxx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    // dA = np.diag(A)
+    // tol = np.min(dA[dA > 0]) * 1e-9
+    Vector<Type, N> d = A.diag();
+    Type tol = d.max();
+    for (size_t k = 0; k < N; k++) {
+        if ((d(k) > 0) && (d(k) < tol)) {
+            tol = d(k);
+        }
+    }
+    tol /= 1000000000;
```
[the remark this answers is not in the document]
Comment (jlecoeur) [reply matrix/PseudoInverse.hxx]: done

[the remark this answers is not in the document]
Comment (jlecoeur) [reply matrix/PseudoInverse.hxx]: done

File: matrix/PseudoInverse.hpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * Courrieu, P. (2008). Fast Computation of Moore-Penrose Inverse Matrices, 8(2), 25–29. http://arxiv.org/abs/0804.4809
+ */
+template<typename Type, size_t M, size_t N>
+Matrix<Type, N, M> geninv(const Matrix<Type, M, N> & G)
+{
+    size_t rank;
+    if (M <= N) {
+        SquareMatrix<Type, M> A = G * G.transpose();
+        SquareMatrix<Type, M> L = fullRankCholesky(A, rank);
```
[the remark this answers is not in the document]
Comment (jlecoeur) [reply matrix/PseudoInverse.hpp]: You are right. I selected this method knowing that it is not the most accurate as a tradeoff between speed and accuracy.
I have tested against Numpy's `pinv` (which is implemented using SVD) on all the matrices generated for the multirotor mixer. `pinv` and `geninv` gave the same results.

File: test/pseudoInverse.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    SquareMatrix<float, 3> A2(data2);
+    SquareMatrix<float, 3> A2_I = inv(A2);
+    SquareMatrix<float, 3> A2_I_check(data2_check);
+    TEST((A2_I - A2_I_check).abs().max() < 1e-5);
+
+    // Test error case with erroneous rank in internal impl functions
+    Matrix<float, 2, 2> L;
+    Matrix<float, 2, 3> GM;
+    Matrix<float, 3, 2> retM_check;
+    Matrix<float, 3, 2> retM = GeninvImpl<float, 2, 3, 0>::genInvUnderdetermined(GM, L, 5);
```
[the remark this answers is not in the document]
Comment (jlecoeur) [reply test/pseudoInverse.cpp]: Thanks! I have no clue what is happening.

File: matrix/PseudoInverse.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkflying) [reply matrix/PseudoInverse.hpp]: > I have tested ... on all the matrices generated for the multirotor mixer

That's probably sufficient then. The norm_min constraint helps a lot in the rank deficient case.

I guess my main worry was about fixed wing / VTOL, but those have issues with actuator saturation at low speed anyway, which would require a non-linear minimization to take the actuator constraints into account.

**Comment by jkflying** [comment]
Ok, I think I got the coverage issue, it was the templating R on 0 in those final lines of the test so it was using the specialization, rather than the generic one.

My only other comment is that this might use a _lot_ of flash space due to having all of the template specializations generated for all of the R from [1 ... M]. I don't know if the performance difference would be significant on our simple in-order no-vectorization microcontrollers, but it would certainly be better from a flash space perspective to have the R as a function parameter rather than a template parameter to get around this.

**Comment by jlecoeur** [comment]
@jkflying Nice, well done!

**Comment by jlecoeur** [comment]
> this might use a lot of flash space due to having all of the template specializations generated

Right. It will generate many specializations. The issue with R a function parameter is that you need to implement everything by hand: you cannot create a square matrix of size R because R is not known at compile-time, so you cannot use the existing `inv()` method, etc.

Here is the flash space usage:
```
make px4_fmu-v5_default bloaty_templates | grep Geninv
   0.4%  7.31Ki    matrix::GeninvImpl<>::genInvUnderdetermined()
     6.8%     512    matrix::GeninvImpl<float, 6u, 16u, 4u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     6.6%     494    matrix::GeninvImpl<float, 6u, 16u, 2u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     6.5%     484    matrix::GeninvImpl<float, 6u, 16u, 3u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     6.4%     478    matrix::GeninvImpl<float, 6u, 16u, 5u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.4%     326    matrix::GeninvImpl<float, 6u, 16u, 1u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.3%     320    matrix::GeninvImpl<float, 6u, 12u, 5u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.2%     318    matrix::GeninvImpl<float, 6u, 12u, 2u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.1%     310    matrix::GeninvImpl<float, 6u, 8u, 2u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.1%     304    matrix::GeninvImpl<float, 3u, 4u, 2u>::genInvUnderdetermined(matrix::Matrix<float, 3u, 4u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
     4.1%     304    matrix::GeninvImpl<float, 4u, 6u, 2u>::genInvUnderdetermined(matrix::Matrix<float, 4u, 6u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     4.0%     302    matrix::GeninvImpl<float, 6u, 12u, 3u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.0%     300    matrix::GeninvImpl<float, 6u, 16u, 6u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 16u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.0%     300    matrix::GeninvImpl<float, 6u, 8u, 5u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.0%     298    matrix::GeninvImpl<float, 6u, 8u, 3u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     4.0%     296    matrix::GeninvImpl<float, 6u, 12u, 4u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.9%     294    matrix::GeninvImpl<float, 4u, 6u, 3u>::genInvUnderdetermined(matrix::Matrix<float, 4u, 6u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     3.9%     292    matrix::GeninvImpl<float, 6u, 8u, 4u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.2%     236    matrix::GeninvImpl<float, 3u, 4u, 1u>::genInvUnderdetermined(matrix::Matrix<float, 3u, 4u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
     3.1%     234    matrix::GeninvImpl<float, 6u, 12u, 1u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.1%     230    matrix::GeninvImpl<float, 6u, 8u, 1u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.0%     228    matrix::GeninvImpl<float, 4u, 6u, 1u>::genInvUnderdetermined(matrix::Matrix<float, 4u, 6u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     2.3%     172    matrix::GeninvImpl<float, 6u, 12u, 6u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 12u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     2.1%     154    matrix::GeninvImpl<float, 4u, 6u, 4u>::genInvUnderdetermined(matrix::Matrix<float, 4u, 6u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     2.1%     154    matrix::GeninvImpl<float, 6u, 8u, 6u>::genInvUnderdetermined(matrix::Matrix<float, 6u, 8u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     2.0%     150    matrix::GeninvImpl<float, 3u, 4u, 3u>::genInvUnderdetermined(matrix::Matrix<float, 3u, 4u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
   0.2%  3.99Ki    matrix::GeninvImpl<>::genInvOverdetermined()
     5.5%     224    matrix::GeninvImpl<float, 6u, 4u, 3u>::genInvOverdetermined(matrix::Matrix<float, 6u, 4u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     5.4%     220    matrix::GeninvImpl<float, 4u, 3u, 2u>::genInvOverdetermined(matrix::Matrix<float, 4u, 3u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
     5.3%     216    matrix::GeninvImpl<float, 6u, 4u, 2u>::genInvOverdetermined(matrix::Matrix<float, 6u, 4u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     4.6%     188    matrix::GeninvImpl<float, 4u, 3u, 1u>::genInvOverdetermined(matrix::Matrix<float, 4u, 3u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
     4.5%     184    matrix::GeninvImpl<float, 6u, 4u, 1u>::genInvOverdetermined(matrix::Matrix<float, 6u, 4u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 12u, 6u, 2u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 12u, 6u, 3u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 12u, 6u, 4u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 12u, 6u, 5u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 12u, 6u, 6u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 16u, 6u, 2u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 16u, 6u, 3u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 16u, 6u, 4u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 16u, 6u, 5u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     156    matrix::GeninvImpl<float, 16u, 6u, 6u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     154    matrix::GeninvImpl<float, 8u, 6u, 2u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     154    matrix::GeninvImpl<float, 8u, 6u, 3u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     154    matrix::GeninvImpl<float, 8u, 6u, 4u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     154    matrix::GeninvImpl<float, 8u, 6u, 5u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.8%     154    matrix::GeninvImpl<float, 8u, 6u, 6u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.7%     150    matrix::GeninvImpl<float, 4u, 3u, 3u>::genInvOverdetermined(matrix::Matrix<float, 4u, 3u> const&, matrix::Matrix<float, 3u, 3u> const&, unsigned int)
     3.7%     150    matrix::GeninvImpl<float, 6u, 4u, 4u>::genInvOverdetermined(matrix::Matrix<float, 6u, 4u> const&, matrix::Matrix<float, 4u, 4u> const&, unsigned int)
     3.5%     142    matrix::GeninvImpl<float, 12u, 6u, 1u>::genInvOverdetermined(matrix::Matrix<float, 12u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.5%     142    matrix::GeninvImpl<float, 16u, 6u, 1u>::genInvOverdetermined(matrix::Matrix<float, 16u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
     3.3%     136    matrix::GeninvImpl<float, 8u, 6u, 1u>::genInvOverdetermined(matrix::Matrix<float, 8u, 6u> const&, matrix::Matrix<float, 6u, 6u> const&, unsigned int)
```

Note that we do not *need* all of these, most are for the tests.

For comparison, here is the flash space usage for `inv()`: 
```
0.3%  4.45Ki    matrix::inv<>()
    17.6%     800    bool matrix::inv<float, 3u>(matrix::SquareMatrix<float, 3u> const&, matrix::SquareMatrix<float, 3u>&)
    14.1%     644    bool matrix::inv<float, 4u>(matrix::SquareMatrix<float, 4u> const&, matrix::SquareMatrix<float, 4u>&)
    13.7%     624    bool matrix::inv<float, 5u>(matrix::SquareMatrix<float, 5u> const&, matrix::SquareMatrix<float, 5u>&)
    13.6%     620    bool matrix::inv<float, 6u>(matrix::SquareMatrix<float, 6u> const&, matrix::SquareMatrix<float, 6u>&)
    12.6%     576    bool matrix::inv<double, 6u>(matrix::SquareMatrix<double, 6u> const&, matrix::SquareMatrix<double, 6u>&)
    12.3%     560    bool matrix::inv<double, 4u>(matrix::SquareMatrix<double, 4u> const&, matrix::SquareMatrix<double, 4u>&)
     9.1%     416    bool matrix::inv<float, 2u>(matrix::SquareMatrix<float, 2u> const&, matrix::SquareMatrix<float, 2u>&)
     2.5%     116    bool matrix::inv<float, 1u>(matrix::SquareMatrix<float, 1u> const&, matrix::SquareMatrix<float, 1u>&)
     0.8%      38    matrix::SquareMatrix<float, 5u> matrix::inv<float, 5u>(matrix::SquareMatrix<float, 5u> const&)
     0.7%      34    matrix::SquareMatrix<float, 2u> matrix::inv<float, 2u>(matrix::SquareMatrix<float, 2u> const&)
     0.7%      34    matrix::SquareMatrix<float, 3u> matrix::inv<float, 3u>(matrix::SquareMatrix<float, 3u> const&)
     0.7%      34    matrix::SquareMatrix<float, 4u> matrix::inv<float, 4u>(matrix::SquareMatrix<float, 4u> const&)
     0.7%      34    matrix::SquareMatrix<float, 6u> matrix::inv<float, 6u>(matrix::SquareMatrix<float, 6u> const&)
     0.6%      28    matrix::SquareMatrix<float, 1u> matrix::inv<float, 1u>(matrix::SquareMatrix<float, 1u> const&)
```

**Comment by jkflying** [comment]
> you cannot create a square matrix of size R because R is not known at compile-time, so you cannot use the existing inv() method, etc

Ok, makes sense. I was hoping one day we'd be able to use a Slice in place of a matrix with full functionality, but until that happens I guess this is the best way forward.

**Comment by jlecoeur** [comment]
@jkflying Thanks for the review and for the help. I added two more tests. Good to go!

**Comment by jlecoeur** [comment]
I am ok with a squash-and-merge, let me know if you prefer that I rebase manually.

**Pull request merged**